README.md
# Micro Fleet - Backend ID Generator library

Belongs to Micro Fleet framework, provides methods to generate database primary key ID.

See more examples and usage guide in unit test.

## INSTALLATION

- Stable version: `npm i @micro-fleet/id-generator`
- Edge (development) version: `npm i git://github.com/gennovative/micro-fleet-id-generator.git`

## DEVELOPMENT

- Install packages in `peerDependencies` section with command `npm i --no-save {package name}@{version}`. Or if you want to use directly neighbor packages, excute `npm run linkPackages`.
- `npm run build` to transpile TypeScript then run unit tests (if any) (equiv. `npm run compile` + `npm run test` (if any)).
- `npm run compile`: To transpile TypeScript into JavaScript.
- `npm run watch`: To transpile without running unit tests, then watch for changes in *.ts files and re-transpile on save.
- `npm run test`: To run unit tests.
  * After tests finish, open file `/coverage/index.html` with a web browser to see the code coverage report which is mapped to TypeScript code.


## RELEASE

- `npm run release`: To transpile and create `app.d.ts` definition file.
- **Note:** Please commit transpiled code in folder `dist` and definition file `app.d.ts` relevant to the TypeScript version.

CHANGELOG.md
## VERSIONS


### 2.3.0
- Sync version with other packages.

### 1.2.0 (Coming soon)
- *IdProvider*: Can fetch from remote source.

### 1.1.3
- Upgraded to new version of `@micro-fleet/common` with breaking change.

### 1.1.2
- Fixed [#1](https://github.com/gennovative/micro-fleet-web/issues/1) - IdProviderAddOn should return bigint type.

### 1.1.1
- Remove script "postinstall" from `package.json`.

### 1.1.0
- `IdGenerator.nextBigInt()` now returns native BigInt.

### 1.0.2
- Fixed node engine version in package.json.
- Refactor to replace `let` with `const`.

### 1.0.1
- Temporarily removed dependency on '@micro-fleet/service-communication'.

### 1.0.0
- *IdProvider*: A service addon that can generates IDs.
- *IdGenerator*: Generates bigint string, shortid, and UUID v4.


src/test/IdProvider.spec.ts
import * as chai from 'chai';
import * as spies from 'chai-spies';

import { IConfigurationProvider, SettingItemDataType, constants, Maybe } from '@micro-fleet/common';
import { IDirectRpcCaller, IRpcResponse } from '@micro-fleet/service-communication';

import { IdProvider } from '../app';

chai.use(spies);

const { SvcSettingKeys: S } = constants;
const expect = chai.expect,
	SVC_SLUG = 'test-svc',
	ID_SVC_ADDR = ['192.168.1.1', '192.168.1.2'];

class MockConfigAddOn implements IConfigurationProvider {

	get enableRemote(): boolean {
		return true;
	}

	public get(key: string, type?: SettingItemDataType): Maybe<number | boolean | string | any[]> {
		switch (key) {
			case S.SERVICE_SLUG: return new Maybe(SVC_SLUG);
			case S.ID_SERVICE_ADDRESSES: return new Maybe(ID_SVC_ADDR);
			default: return new Maybe;
		}
	}

	public deadLetter(): Promise<void> {
		return Promise.resolve();
	}

	public fetch(): Promise<boolean> {
		return Promise.resolve(true);
	}

	public init(): Promise<void> {
		return Promise.resolve();
	}

	public dispose(): Promise<void> {
		return Promise.resolve();
	}

	public onUpdate(listener: (delta: string[]) => void) {
	}
}

class MockDirectRpcCaller implements IDirectRpcCaller {
	public baseAddress: string;
	public name: string;
	public timeout: number;

	public init(params?: any): any {
	}

	public dispose(): Promise<void> {
		return Promise.resolve();
	}

	public onError(handler: (err: any) => void): void {
	}

	public call(moduleName: string, action: string, params?: any): Promise<IRpcResponse> {
		return Promise.resolve(null);
	}

}

let idProd: IdProvider;

describe('IdProvider', () => {

	beforeEach(() => {
		idProd = new IdProvider(new MockConfigAddOn(), new MockDirectRpcCaller());
		return idProd.init();
	});

	describe('init', () => {
		it('should init RPC caller', async () => {
			// Arrange
			let idProd = new IdProvider(new MockConfigAddOn(), new MockDirectRpcCaller());

			// Act
			await idProd.init();

			// Assert
			expect(idProd['_rpcCaller'].name).to.equal(SVC_SLUG);
			expect(idProd['_addresses']).to.equal(ID_SVC_ADDR);
		});
	}); // describe 'init'

	describe('dispose', () => {
		it('should release all resources', async () => {
			// Arrange
			let disposeSpy = chai.spy.on(idProd['_rpcCaller'], 'dispose');

			// Act
			await idProd.dispose();

			// Assert
			
			expect(disposeSpy).to.be.called;
		});
	}); // describe 'dispose'

	describe('nextBigInt', () => {
		it('should return a big int string', async () => {
			// Act
			let id = idProd.nextBigInt();

			// Assert
			console.log('Big Int: ', id);
			expect(id).to.be.string;
			expect(id).not.to.be.empty;
		});
	}); // describe 'nextBigInt'

	describe('nextShortId', () => {
		it('should return a short ID string', async () => {
			// Act
			let id = idProd.nextShortId();

			// Assert
			console.log('Short ID: ', id);
			expect(id).to.be.string;
			expect(id).not.to.be.empty;
		});
	}); // describe 'nextShortId'

	describe('nextUuidv4', () => {
		it('should return a version-4 UUID string', async () => {
			// Act
			let id = idProd.nextUuidv4();

			// Assert
			console.log('UUID V6: ', id);
			expect(id).to.be.string;
			expect(id).not.to.be.empty;
		});
	}); // describe 'nextUuidv4'
});

src/test/IdGenerator.spec.ts
import * as chai from 'chai'
import * as spies from 'chai-spies'

import { IdGenerator } from '../app'


chai.use(spies)
const expect = chai.expect

describe('IdGenerator', () => {
    describe('nextBigInt', () => {
        it('Should return bigint type', () => {
            const idGen = new IdGenerator()
            const result = idGen.nextBigInt()
            console.log(result)
            expect(typeof result).equal('bigint')
        })

        it('Should generate different numbers on each call', () => {
            const idGen = new IdGenerator()
            expect(idGen.nextBigInt()).not.to.equal(idGen.nextBigInt())
        })
    }) // END describe 'nextBigInt'

    describe('nextShortId', () => {
        it('Should return a string', () => {
            const idGen = new IdGenerator({ worker: 2 })
            const result = idGen.nextShortId()
            console.log(result)
            expect(result).not.to.be.string
            expect(result.length >= 7 && result.length <= 14).to.be.true
        })

        it('Should generate different numbers on each call', () => {
            const idGen = new IdGenerator({ worker: 2 })
            expect(idGen.nextShortId()).not.to.equal(idGen.nextShortId())
        })
    }) // END describe 'nextShortId'

    describe('nextUuidv4', () => {
        it('Should return a string', () => {
            const idGen = new IdGenerator()
            const result = idGen.nextUuidv4() // b34c842e-7eca-4504-9999-c7ac3523649c
            console.log(result)
            expect(result).to.be.string
            // UUID v4 should have 37 characters in length
            expect(result.length).to.equal(36)
            // UUID v4 should have third part starting with "4"
            expect(result.split('-')[2].charAt(0)).to.equal('4')
        })

        it('Should generate different numbers on each call', () => {
            const idGen = new IdGenerator()
            expect(idGen.nextUuidv4()).not.to.equal(idGen.nextUuidv4())
        })
    }) // END describe 'nextUuidv4'

    describe('wrapBigInt', () => {
        it('Should call next() internally if no argument is given.', () => {
            // Arrange
            const idGen = new IdGenerator(),
                nextSpy = chai.spy.on(idGen, 'nextBigInt')

            // Act
            idGen.wrapBigInt()

            // Assert
            expect(nextSpy).to.be.called.once
        })

        it('Should wrap a numeric string.', () => {
            // Arrange
            const idGen = new IdGenerator(),
                input = '12345'

            // Act
            const bigId = idGen.wrapBigInt(input)

            // Assert
            expect(typeof bigId).to.equal('bigint')
            expect(bigId).to.equal(BigInt(input))
        })

        it('Should wrap a Buffer.', () => {
            // Arrange
            const idGen = new IdGenerator()
            const buf = Buffer.allocUnsafe(8)
            buf.fill(0)
            buf.writeUInt8(0x5A, 0)
            buf.writeUInt8(0x6D, 1)
            buf.writeUInt8(0x1C, 2)
            buf.writeUInt8(0x38, 3)
            buf.writeUInt8(0x2C, 4)

            // Act
            const bigId = idGen.wrapBigInt(buf)

            // Assert
            expect(typeof bigId).to.equal('bigint')
            expect(bigId === BigInt('6515895263456919552')).to.be.true
        })
    }) // END describe 'wrapBigInt'
})


src/test/register-addon.spec.ts
import * as chai from 'chai'
import * as spies from 'chai-spies'
chai.use(spies)
const expect = chai.expect
import { IDependencyContainer, DependencyContainer, serviceContext } from '@micro-fleet/common'

import { registerIdAddOn, IdProviderAddOn, Types as T } from '../app'


describe('registerIdAddOn', function () {
    // this.timeout(60000) // For debuging

    let depCon: IDependencyContainer

    beforeEach(() => {
        depCon = new DependencyContainer()
        serviceContext.setDependencyContainer(depCon)
    })

    afterEach(() => {
        depCon.dispose()
        depCon = null
    })

    it('Should register dependencies if not already', () => {
        // Act
        registerIdAddOn()

        // Assert
        expect(depCon.isBound(T.ID_PROVIDER)).to.be.true
    })

    it('Should not register dependencies if already registered', () => {
        // Arrange
        depCon.bind<IdProviderAddOn>(T.ID_PROVIDER, IdProviderAddOn)
        chai.spy.on(depCon, 'bind')

        // Act
        registerIdAddOn()

        // Assert
        expect(depCon.bind).not.to.be.called
    })
}) // describe


src/app/IdProvider.ts
import { IConfigurationProvider, Types as ConT, constants,
	injectable, inject, Guard } from '@micro-fleet/common';
import { IDirectRpcCaller, IRpcResponse, Types as ComT } from '@micro-fleet/service-communication';

import { IdGenerator } from './IdGenerator';

const { SvcSettingKeys: SvcS, ModuleNames: M, ActionNames: A } = constants;


@injectable()
export class IdProvider implements IServiceAddOn {

	private static CACHE_SIZE = 10;

	private _addresses: string[];

	// TODO: Will implement remote ID generation later
	private _idGen: IdGenerator;

	constructor(
		@inject(ConT.CONFIG_PROVIDER) private _configProvider: IConfigurationProvider,
		@inject(ComT.DIRECT_RPC_CALLER) private _rpcCaller: IDirectRpcCaller
	) {
		Guard.assertArgDefined('_configProvider', _configProvider);
		Guard.assertArgDefined('_rpcCaller', _rpcCaller);
		this._idGen = new IdGenerator();
	}

	/**
	 * @see IServiceAddOn.init
	 */
	public init(): Promise<void> {
		this._rpcCaller.name = this._configProvider.get(SvcS.SERVICE_SLUG).value as string;
		this._addresses = this._configProvider.get(SvcS.ID_SERVICE_ADDRESSES).value as string[];
		return Promise.resolve();
	}

	/**
	 * @see IServiceAddOn.deadLetter
	 */
	public deadLetter(): Promise<void> {
		return Promise.resolve();
	}

	/**
	 * @see IServiceAddOn.dispose
	 */
	public dispose(): Promise<void> {
		return this._rpcCaller.dispose();
	}

	public async fetch(): Promise<void> {
		for (let addr of this._addresses) {
			if (await this.attempFetch(addr)) {
				return;
			}
		}
	}

	public nextBigInt(): string {
		return this._idGen.nextBigInt().toString();
	}

	public nextShortId(): string {
		return this._idGen.nextShortId();
	}

	public nextUuidv4(): string {
		return this._idGen.nextUuidv4();
	}


	private async attempFetch(address: string): Promise<boolean> {
		this._rpcCaller.baseAddress = address;
		return this._rpcCaller.call(M.ID_GEN, A.NEXT_BIG_INT, {
				service: this._rpcCaller.name,
				count: IdProvider.CACHE_SIZE
			})
			.then((res: IRpcResponse) => {
				// resolve(<any>res.data);
				return true;
			});
	}
}

src/app/IdGenerator.ts
import * as FlakeId from 'flake-idgen'
import * as shortid from 'shortid'
import uuidv4 = require('uuid/v4')


export type BigIdOptions = {

    /**
     * Datacenter identifier. It can have values from 0 to 31 (5 bits).
     */
    datacenter?: number,

    /**
     * Worker identifier. It can have values from 0 to 31 (5 bits).
     */
    worker?: number,

    /**
     * Generator identifier. It can have values from 0 to 1023 (10 bits).
     * It can be provided instead of `datacenter` and `worker` identifiers.
     */
    id?: number,

    /**
     * Number used to reduce value of a generated timestamp.
     * Note that this number should not exceed number of milliseconds elapsed
     * since 1 January 1970 00:00:00 UTC (value of `Date.now()`).
     * It can be used to generate smaller ids.
     */
    epoch?: number,
}

export type Int64 = {
    toNumber(): number,
    toJSON(): number,
    toString(radix?: number): string,
    toBuffer(raw?: boolean): Buffer,
}

/**
 * Provides methods to generate bigint ID
 */
export class IdGenerator {

    private _generator: FlakeId

    constructor(options?: BigIdOptions) {
        this._generator = new FlakeId(options)
        if (options && options.worker) {
            shortid.worker(options.worker)
        }
    }

    /**
     * Generates a new big int ID.
     */
    public nextBigInt(): bigint {
        return bufferToBigInt(this._generator.next())
    }

    /**
     * Generates a 7-character UID.
     */
    public nextShortId(): string {
        return shortid.generate()
    }

    /**
     * Generates a version-4 UUID.
     */
    public nextUuidv4(): string {
        return uuidv4()
    }

    // public wrapBigInt(value: string): Int64
    // public wrapBigInt(buf: Buffer): Int64
    // public wrapBigInt(value?: number): Int64

    /**
     * Parses input value into bigint type.
     * @param value The value to be wrapped. If not given, the behavior is same with `next()`.
     */
    public wrapBigInt(value?: string | number | Buffer): bigint {
        if (value == null) {
            return this.nextBigInt()
        }
        switch (typeof value) {
            case 'bigint':
                return value
            case 'string':
            case 'number':
                return BigInt(value)
            default:
                if (value instanceof Buffer) {
                    return bufferToBigInt(value)
                }
                return BigInt(value)
        }
    }
}

/**
 * Converts a buffer to bigint
 */
function bufferToBigInt(buffer: Buffer): bigint {
    // Eg: Buffer <5A, 6D, 1C, 38, 2C, 00, 00, 00>
    // value = (0x5A << 7) + (0x6D << 6) ....
    const BITS_PER_BYTE = BigInt(8)
    const result = [...buffer].reduce((prev, cur, idx, arr) => {
        const reversedIdx = BigInt(arr.length - 1 - idx)
        return prev + (BigInt(cur) << (reversedIdx * BITS_PER_BYTE))
    }, BigInt(0))

    return result
}


src/app/IdProviderAddOn.ts
import { IServiceAddOn, IConfigurationProvider, constants,
    SettingItemDataType, decorators as d, Types as T,
} from '@micro-fleet/common'

import { IdGenerator, BigIdOptions } from './IdGenerator'


const {
    IdGenerator: C,
} = constants


export interface IIdProvider {
    /**
     * Generates a chronologically sequential native bigint with Snowflake algorithm.
     */
    nextBigInt(): bigint

    /**
     * Generates a random string of length from 7 to 14 url-friendly characters.
     */
    nextShortId(): string

    /**
     * Generates a v4 universally unique identifier.
     */
    nextUuidv4(): string
}

@d.injectable()
export class IdProviderAddOn implements IIdProvider, IServiceAddOn {

    public readonly name: string = 'IdProviderAddOn'

    // private static CACHE_SIZE = 10

    // private _addresses: string[]

    // TODO: Will implement remote ID generation later
    private _idGen: IdGenerator


    // @lazyInject(ComT.DIRECT_RPC_CALLER)
    // @lazyInject('service-communication.IDirectRpcCaller')
    // private _rpcCaller: IDirectRpcCaller

    constructor(
        @d.inject(T.CONFIG_PROVIDER) private _config: IConfigurationProvider
    ) {
    }

    /**
     * @see IServiceAddOn.init
     */
    public init(): Promise<void> {
        const opts: BigIdOptions = {}
        const getCfg = this._config.get.bind(this._config)
        getCfg(C.ID_DATACENTER, SettingItemDataType.Number).map(val => opts.datacenter = val)
        getCfg(C.ID_WORKER, SettingItemDataType.Number).map(val => opts.worker = val)
        getCfg(C.ID_EPOCH, SettingItemDataType.Number).map(val => opts.epoch = val)
        this._idGen = new IdGenerator(opts)
        return Promise.resolve()
    }

    /**
     * @see IServiceAddOn.deadLetter
     */
    public deadLetter(): Promise<void> {
        return Promise.resolve()
    }

    /**
     * @see IServiceAddOn.dispose
     */
    public dispose(): Promise<void> {
        return Promise.resolve()
        // if (!this._rpcCaller) {
        //     return Promise.resolve()
        // }
        // return this._rpcCaller.dispose()
    }

    // public async fetch(): Promise<void> {
    //     this._rpcCaller.name = this._configProvider.get(SvcS.SERVICE_SLUG).value as string
    //     this._addresses = this._configProvider.get(SvcS.ID_SERVICE_ADDRESSES).value as string[]

    //     // tslint:disable-next-line:prefer-const
    //     for (let addr of this._addresses) {
    //         if (await this.attempFetch(addr)) {
    //             return
    //         }
    //     }
    // }

    /**
     * @see IIdProvider.nextBigInt
     */
    public nextBigInt(): bigint {
        return this._idGen.nextBigInt()
    }


    /**
     * @see IIdProvider.nextShortId
     */
    public nextShortId(): string {
        return this._idGen.nextShortId()
    }

    /**
     * @see IIdProvider.nextUuidv4
     */
    public nextUuidv4(): string {
        return this._idGen.nextUuidv4()
    }


    // private async attempFetch(address: string): Promise<boolean> {
    //     this._rpcCaller.baseAddress = address
    //     return this._rpcCaller.call(M.ID_GEN, A.NEXT_BIG_INT, {
    //             service: this._rpcCaller.name,
    //             count: IdProviderAddOn.CACHE_SIZE,
    //         })
    //         .then((res: IRpcResponse) => {
    //             // resolve(<any>res.data)
    //             return true
    //         })
    // }
}


typings/app.d.ts
/// <reference path="./global.d.ts" />
declare module '@micro-fleet/id-generator/dist/app/IdGenerator' {
    /// <reference types="node" />
    export type BigIdOptions = {
        /**
         * Datacenter identifier. It can have values from 0 to 31 (5 bits).
         */
        datacenter?: number;
        /**
         * Worker identifier. It can have values from 0 to 31 (5 bits).
         */
        worker?: number;
        /**
         * Generator identifier. It can have values from 0 to 1023 (10 bits).
         * It can be provided instead of `datacenter` and `worker` identifiers.
         */
        id?: number;
        /**
         * Number used to reduce value of a generated timestamp.
         * Note that this number should not exceed number of milliseconds elapsed
         * since 1 January 1970 00:00:00 UTC (value of `Date.now()`).
         * It can be used to generate smaller ids.
         */
        epoch?: number;
    };
    export type Int64 = {
        toNumber(): number;
        toJSON(): number;
        toString(radix?: number): string;
        toBuffer(raw?: boolean): Buffer;
    };
    /**
     * Provides methods to generate bigint ID
     */
    export class IdGenerator {
                constructor(options?: BigIdOptions);
        /**
         * Generates a new big int ID.
         */
        nextBigInt(): bigint;
        /**
         * Generates a 7-character UID.
         */
        nextShortId(): string;
        /**
         * Generates a version-4 UUID.
         */
        nextUuidv4(): string;
        /**
         * Parses input value into bigint type.
         * @param value The value to be wrapped. If not given, the behavior is same with `next()`.
         */
        wrapBigInt(value?: string | number | Buffer): bigint;
    }

}
declare module '@micro-fleet/id-generator/dist/app/IdProviderAddOn' {
    import { IServiceAddOn, IConfigurationProvider } from '@micro-fleet/common';
    export interface IIdProvider {
        /**
         * Generates a chronologically sequential native bigint with Snowflake algorithm.
         */
        nextBigInt(): bigint;
        /**
         * Generates a random string of length from 7 to 14 url-friendly characters.
         */
        nextShortId(): string;
        /**
         * Generates a v4 universally unique identifier.
         */
        nextUuidv4(): string;
    }
    export class IdProviderAddOn implements IIdProvider, IServiceAddOn {
                readonly name: string;
                constructor(_config: IConfigurationProvider);
        /**
         * @see IServiceAddOn.init
         */
        init(): Promise<void>;
        /**
         * @see IServiceAddOn.deadLetter
         */
        deadLetter(): Promise<void>;
        /**
         * @see IServiceAddOn.dispose
         */
        dispose(): Promise<void>;
        /**
         * @see IIdProvider.nextBigInt
         */
        nextBigInt(): bigint;
        /**
         * @see IIdProvider.nextShortId
         */
        nextShortId(): string;
        /**
         * @see IIdProvider.nextUuidv4
         */
        nextUuidv4(): string;
    }

}
declare module '@micro-fleet/id-generator/dist/app/Types' {
    export class Types {
        static readonly ID_PROVIDER = "id-generator.IdProvider";
    }

}
declare module '@micro-fleet/id-generator/dist/app/constants' {
    export enum IdSettingKeys {
        /**
         * Array of addresses to fetch IDs.
         * Data type: string[]
         */
        ID_SERVICE_ADDRESSES = "id_service_addresses"
    }

}
declare module '@micro-fleet/id-generator/dist/app/register-addon' {
    import { IdProviderAddOn } from '@micro-fleet/id-generator/dist/app/IdProviderAddOn';
    export function registerIdAddOn(): IdProviderAddOn;

}
declare module '@micro-fleet/id-generator' {
    export * from '@micro-fleet/id-generator/dist/app/constants';
    export * from '@micro-fleet/id-generator/dist/app/IdGenerator';
    export * from '@micro-fleet/id-generator/dist/app/IdProviderAddOn';
    export * from '@micro-fleet/id-generator/dist/app/register-addon';
    export * from '@micro-fleet/id-generator/dist/app/Types';

}
